2D slice definition with NEO3.0 compiler

I found out this code panics at compile time in latest master branch (0.90.0-pre-529-gf75d888f) but runs correctly in master-2.x branch (0.75.1-pre-80-g3ccf19fc)

```go
package foo
func Main() int {
	a := [][]int{{0,1},{2,3}}
	b := a[0]
	return b[1]
}
```

If I change definition to `a := [][]int{[]int{0,1},[]int{2,3}}` there is no panic.

Panic trace

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x876d88]

goroutine 55 [running]:
testing.tRunner.func1.1(0xac4960, 0x1087d10)
	/snap/go/current/src/testing/testing.go:940 +0x2f5
testing.tRunner.func1(0xc0002d4900)
	/snap/go/current/src/testing/testing.go:943 +0x3f9
panic(0xac4960, 0x1087d10)
	/snap/go/current/src/runtime/panic.go:969 +0x166
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).Visit(0xc0002bea50, 0xc71f40, 0xc0002cb680, 0xc78780, 0xc0002d1d40)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:597 +0x1138
go/ast.Walk(0xc6cc00, 0xc0002bea50, 0xc71f40, 0xc0002cb680)
	/snap/go/current/src/go/ast/walk.go:52 +0x66
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).Visit(0xc0002bea50, 0xc71f40, 0xc0002cb6c0, 0x0, 0xc0002d8500)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:614 +0x1297
go/ast.Walk(0xc6cc00, 0xc0002bea50, 0xc71f40, 0xc0002cb6c0)
	/snap/go/current/src/go/ast/walk.go:52 +0x66
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).Visit(0xc0002bea50, 0xc71bc0, 0xc0002cb700, 0x0, 0xc6dd60)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:413 +0x3aa7
go/ast.Walk(0xc6cc00, 0xc0002bea50, 0xc71bc0, 0xc0002cb700)
	/snap/go/current/src/go/ast/walk.go:52 +0x66
go/ast.walkStmtList(0xc6cc00, 0xc0002bea50, 0xc0002cb780, 0x3, 0x4)
	/snap/go/current/src/go/ast/walk.go:32 +0x9e
go/ast.Walk(0xc6cc00, 0xc0002bea50, 0xc71d40, 0xc0002d1d70)
	/snap/go/current/src/go/ast/walk.go:224 +0x1adf
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).convertFuncDecl(0xc0002bea50, 0xc72140, 0xc0000a1700, 0xc0002d1da0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:335 +0x3ec
github.com/nspcc-dev/neo-go/pkg/compiler.(*codegen).compile(0xc0002bea50, 0xc0002de440, 0xc0002be9a0, 0xc0002be9a0, 0x0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:1386 +0x1ec
github.com/nspcc-dev/neo-go/pkg/compiler.CodeGen(0xc0002de440, 0xc0002de0e0, 0xc0002de440, 0x0, 0x0, 0xc0002d1b01, 0xc0002de0e0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/codegen.go:1437 +0x88
github.com/nspcc-dev/neo-go/pkg/compiler.CompileWithDebugInfo(0xc6d480, 0xc0002de0e0, 0xd9af48, 0xf, 0x0, 0x0, 0x0, 0x5151c0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/compiler.go:78 +0x90
github.com/nspcc-dev/neo-go/pkg/compiler.Compile(...)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/compiler.go:64
github.com/nspcc-dev/neo-go/pkg/compiler_test.vmAndCompileInterop(0xc0002d4900, 0xba1902, 0x62, 0x37e, 0x5151c0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/vm_test.go:68 +0x110
github.com/nspcc-dev/neo-go/pkg/compiler_test.vmAndCompile(...)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/vm_test.go:58
github.com/nspcc-dev/neo-go/pkg/compiler_test.eval(0xc0002d4900, 0xba1902, 0x62, 0xb73980, 0xc00009afc0)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/vm_test.go:35 +0x49
github.com/nspcc-dev/neo-go/pkg/compiler_test.runTestCases.func1(0xc0002d4900)
	/home/alexvanin/Project/Neo/neo-go/pkg/compiler/vm_test.go:24 +0x53
testing.tRunner(0xc0002d4900, 0xc0002d8380)
	/snap/go/current/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
	/snap/go/current/src/testing/testing.go:1042 +0x357
```

--- a/pkg/compiler/codegen.go
+++ b/pkg/compiler/codegen.go
@@ -594,15 +594,11 @@ func (c *codegen) Visit(node ast.Node) ast.Visitor {
 		return nil
 
 	case *ast.CompositeLit:
-		typ := c.typeOf(n.Type).Underlying()
-		switch n.Type.(type) {
-		case *ast.Ident, *ast.SelectorExpr, *ast.MapType:
-			switch typ.(type) {
-			case *types.Struct:
-				c.convertStruct(n)
-			case *types.Map:
-				c.convertMap(n)
-			}
+		switch typ := c.typeOf(n).Underlying().(type) {
+		case *types.Struct:
+			c.convertStruct(n)
+		case *types.Map:
+			c.convertMap(n)
 		default:
 			ln := len(n.Elts)
 			// ByteArrays needs a different approach than normal arrays.

--- a/pkg/compiler/codegen.go
+++ b/pkg/compiler/codegen.go
@@ -1253,11 +1253,24 @@ func (c *codegen) convertStruct(lit *ast.CompositeLit) {
 	emit.Int(c.prog.BinWriter, int64(strct.NumFields()))
 	emit.Opcode(c.prog.BinWriter, opcode.NEWSTRUCT)
 
+	keyedLit := len(lit.Elts) > 0
+	if keyedLit {
+		_, ok := lit.Elts[0].(*ast.KeyValueExpr)
+		keyedLit = keyedLit && ok
+	}
 	// We need to locally store all the fields, even if they are not initialized.
 	// We will initialize all fields to their "zero" value.
 	for i := 0; i < strct.NumFields(); i++ {
 		sField := strct.Field(i)
 		fieldAdded := false
+
+		if !keyedLit {
+			emit.Opcode(c.prog.BinWriter, opcode.DUP)
+			emit.Int(c.prog.BinWriter, int64(i))
+			ast.Walk(c, lit.Elts[i])
+			emit.Opcode(c.prog.BinWriter, opcode.SETITEM)
+			continue
+		}
 
 		// Fields initialized by the program.
 		for _, field := range lit.Elts {

--- a/pkg/compiler/slice_test.go
+++ b/pkg/compiler/slice_test.go
@@ -222,6 +222,27 @@ var sliceTestCases = []testCase{
 		}`,
 		big.NewInt(42),
 	},
+	{
+		"nested slice omitted type (slice)",
+		`package foo
+		func Main() int {
+			a := [][]int{{1, 2}, {3, 4}}
+			a[1][0] = 42
+			return a[1][0]
+		}`,
+		big.NewInt(42),
+	},
+	{
+		"nested slice omitted type (struct)",
+		`package foo
+		type pair struct { a, b int }
+		func Main() int {
+			a := []pair{{a: 1, b: 2}, {a: 3, b: 4}}
+			a[1].a = 42
+			return a[1].a
+		}`,
+		big.NewInt(42),
+	},
 }
 
 func TestSliceOperations(t *testing.T) {

--- a/pkg/compiler/struct_test.go
+++ b/pkg/compiler/struct_test.go
@@ -394,6 +394,17 @@ var structTestCases = []testCase{
 		}`,
 		big.NewInt(42),
 	},
+	{
+		"omit field names",
+		`package foo
+		type pair struct { a, b int }
+		func Main() int {
+			p := pair{1, 2}
+			x := p.a * 10
+			return x + p.b
+		}`,
+		big.NewInt(12),
+	},
 }
 
 func TestStructs(t *testing.T) {

